transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -323,160 +349,239 @@ private String generateName(ChannelHandler handler) {
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public final <T extends ChannelHandler> T replace(
             Class<T> oldHandlerType, String newName, ChannelHandler newHandler) {
-        return (T) replace(getContextOrDie(oldHandlerType), newName, newHandler);
+        return (T) replace(ctx -> oldHandlerType.isAssignableFrom(ctx.handler().getClass()), newName, newHandler);
     }
 
     private ChannelHandler replace(
-            final AbstractChannelHandlerContext ctx, String newName, ChannelHandler newHandler) {
-        assert ctx != head && ctx != tail;
+            Predicate<AbstractChannelHandlerContext> predicate, String newName, ChannelHandler newHandler) {
+        checkMultiplicity(newHandler);
+
+        if (newName == null) {
+            newName = generateName(newHandler);
+        }
+        AbstractChannelHandlerContext oldCtx;
+        AbstractChannelHandlerContext newCtx = newContext(newName, newHandler);
+        final int idx;
+        synchronized (handlers) {
+            idx = findCtxIdx(predicate);
+            if (idx == -1) {
+                throw new NoSuchElementException();
+            }
+            oldCtx = handlers.get(idx);
+            assert oldCtx != head && oldCtx != tail && oldCtx != null;
 
-        final AbstractChannelHandlerContext newCtx;
-        synchronized (this) {
-            checkMultiplicity(newHandler);
-            if (newName == null) {
-                newName = generateName(newHandler);
-            } else {
-                boolean sameName = ctx.name().equals(newName);
-                if (!sameName) {
-                    checkDuplicateName(newName);
+            if (!oldCtx.name().equals(newName)) {
+                if (context(newName) != null) {
+                    throw new IllegalArgumentException("Duplicate handler name: " + newName);
                 }
             }
+            AbstractChannelHandlerContext removed = handlers.remove(idx);
+            assert removed != null;
+            handlers.add(idx, newCtx);
+        }
 
-            newCtx = newContext(ctx.executor, newName, newHandler);
-
-            replace0(ctx, newCtx);
-
-            EventExecutor executor = ctx.executor();
-            if (!executor.inEventLoop()) {
-                executor.execute(() -> {
-                    // Invoke newHandler.handlerAdded() first (i.e. before oldHandler.handlerRemoved() is invoked)
-                    // because callHandlerRemoved() will trigger channelRead() or flush() on newHandler and
-                    // those event handlers must be called after handlerAdded().
-                    callHandlerAdded0(newCtx);
-                    callHandlerRemoved0(ctx);
-                });
-                return ctx.handler();
+        EventExecutor executor = executor();
+        if (executor.inEventLoop()) {
+            replace0(oldCtx, newCtx);
+        } else {
+            try {
```

Comment:
I think this still needs to be moved inside the sync block? The `final int idx;` declaration can then also be too